### PR TITLE
add a boxed CPU fallback kernel

### PR DESCRIPTION
**Pull request by bdhirsh**
This PR replaces the existing code-generated CPU fallback kernels that XLA uses with a single boxed CPU fallback.

Current state: there are a couple different design ideas that I want to point out, but the logic for the actually kernel is mostly done and passing tests.

### Design

To preface, I'm not 100% tied to the current design and I'm putting the PR up now for opinions and totally open to alternatives, some of which I listed below. Actually after writing this description, I'm leaning toward the following changes:
* Confirm whether or not we can remove all C++ logging info directly in the yaml.

**Current Design**

All of the CPU fallback codegen is deleted. In its place, XLA (and other external backends, later) can choose to opt into a CPU fallback by adding the following code in a C++ file. I have an corresponding [xla-side PR with the xla changes](https://github.com/pytorch/xla/pull/2945/files#diff-1a005c10039f0cb11130a3b740f5de716d2f10acaea121017016025861886798R1).

There's no actual requirement to split up the code into a .h and .cpp file, but that's necessary in the XLA case because they sometimes need to call the fallback directly from their handcrafted kernels.

```
// xla_cpu_fallback.h
#include <ATen/native/CPUFallback.h>
...
void xla_cpu_fallback(const c10::OperatorHandle& op, torch::jit::Stack* stack);
...
```
```
// xla_cpu_fallback.cpp
#include "torch_xla/csrc/aten_cpu_fallback.h"
...
void xla_cpu_fallback(const c10::OperatorHandle& op, torch::jit::Stack* stack) {
  // Do custom logging here
  ...
  // Call the actual boxed CPU fallback.
  at::native::cpu_fallback(op, stack);
}

TORCH_LIBRARY_IMPL(_, XLA, m) {
  m.fallback(torch::CppFunction::makeFromBoxedFunction<&xla_cpu_fallback>());
}
```

Now that the fallback is exposed in the backend, they can call it directly. Doing so requires converting from an unboxed to a boxed context, which we provide a utility function before. E.g.:
```
#include <ATen/native/CPUFallback.h>

at::Tensor addmm(const at::Tensor& self,const at::Tensor& mat1,const at::Tensor& mat2,const at::Scalar& beta,const at::Scalar& alpha) {
  ....
  if (...call_fallback...) {
    return at::native::call_fallback_fn<&xla_cpu_fallback, decltype(at::addmm)>::call("aten::addmm", self, mat1, mat2, beta, alpha);
  }
  ...
}
```

That `decltype(at::addmm)` logic isn't actually used everywhere in the xla-side PR yet, since you hit issues with overloads. I could use it everywhere once #58092 lands.

**Alternatives: The API for calling the CPU fallback directly is ugly, can we make it nicer?**
We could change the api to use `at::redispatch`, which would make it look something like this:
```
at::Tensor addmm(const at::Tensor& self,const at::Tensor& mat1,const at::Tensor& mat2,const at::Scalar& beta,const at::Scalar& alpha) {
  ....
  if (...call_fallback...) {
    return at::redispatch::addmm(c10::DispatchKeySet(c10::DispatchKey::CPUFallback), self, mat1, mat2, beta, alpha);
  }
  ...
}
```
Which definitely feels cleaner, but also requires adding a new DispatchKey just for this use case. Conditionally calling the CPU fallback doesn't sound like a hugely important use case, so I don't know if giving up one of our 64 dispatch key slots is worth the API improvement. Totally open to other opinions though!

Another more mild improvement that would avoid having to pass operator string names (including overloads) around would be to codegen (yet another) namespaced API. Something like this:
```
at::Tensor addmm(const at::Tensor& self,const at::Tensor& mat1,const at::Tensor& mat2,const at::Scalar& beta,const at::Scalar& alpha) {
  ....
  if (...call_fallback...) {
    return at::fallback::addmm<&xla_cpu_fallback>(self, mat1, mat2, beta, alpha);
  }
  ...
}
```

Writing that out actually I actually like it more (I think it'll let us get rid of `decltype(...)`). Maybe that is nice enough to warrant a new codegen API - I haven't tried adding that yet, but if people like it I'm happy to try it out.

**More alternatives**
The current design also involves the backend manually writing and registering the boxed fallback themselves, but an alternative would be for us to do it in codegen too: they would just need to pass in all of the C++ logging that they want done in the fallback, directly through the yaml. The main downsides:
* Backend code that wants to call the fallback needs to abide by whatever convention our codegen uses to name the generated boxed fallback.
* Passing custom C++ logging through yaml is just more fragile: right now xla uses an `iostream` to log each tensor arg in the operator, so we'd have to either force other backends into the same convention or figure something else out later.

To be fair, we actually already do that: XLA has custom per-tensor-arg logging for all of the generated `out` wrappers in the codegen, which we do by passing their C++ logging info through the yaml. This seems unnecessary though, since `out` wrappers just call into a functional kernel, which is hand written with its own custom logging. So my take is: try to remove custom C++ logging from the yaml, and if it turns out to be really necessary, then we may as well take advantage of that to codegen the fallback.

### Performance impact

While ops that fall back to CPU aren't exactly hot path, we probably don't want to use a boxed fallback if it turns out to be an absolute perf killer.

I ran my benchmarks using callgrind, benchmarking both `at::add` and `at::add_out` run on XLA. My callgrind benchmark for `at::add` can be found here (the add_out benchmark looks basically the same): https://www.internalfb.com/phabricator/paste/view/P415418587. I created the benchmark by hacking the existing xla C++ test build scripts and throwing in a reference to callgrind.

I also attached the full callgrind output for each benchmark; the full output is actually pretty noise and hard to parse, but I focused on everything underneath the `at::add()` call in the output, which was much more stable. My guess is that it's due to some heavyweight async startup processing that xla does.

`at::add`:
before: 88,505,130 instructions. Full output: https://www.internalfb.com/phabricator/paste/view/P415421001
after: 102,185,654 instructions. Full output: https://www.internalfb.com/phabricator/paste/view/P415421273
delta: ~15.5% increase

`at::add_out`:
before: 63,897,395 instructions. Full output: https://www.internalfb.com/intern/everpaste/?handle=GBrrKwtAPlix9wUEAOZtrFXpdO5UbsIXAAAz
after: 73,170,346 instructions. Full output: https://www.internalfb.com/phabricator/paste/view/P415423227
delta: ~14.5% increase

High level takeaway: A framework overhead increase of 10-20% doesn't seem too horrible for the CPU fallback use case.

For structured, functional ops that requires a CPU fallback, we're actually in an unfortunate situation: we're doing even more work than necessary. Our codegen automatically creates a `CompositeExplicitAutograd` kernel which calls into the `out` operator. So the extra work that we end up doing is:
* An extra dispatcher hop: (at::add -> CompositeExplicitAutograd -> CPUFallback -> at::native::add) instead of (at::add -> CPUFallback -> at::native::add)
* An unnecessary tensor allocation (the CompositeExplicitAutograd kernel uses at::empty() to create an output tensor, which is immediately overwritten by the CPU fallback)
* An unnecessary meta() call (the CompositeExplicitAutograd kernel calls it to create the output tensor, but we call it again in the CPU kernel).
* unboxing->boxing->unboxing logic (this is the only strictly required piece)

There are definitely ways to avoid the unnecessary work explained above: one would be to give the boxed fallback higher priority than composite keys (there's [an issue for it here](https://github.com/pytorch/pytorch/issues/55104)), and codegen fallthroughs for all composite ops. It'll require more infra to set up, so I see it as more of a perf knob that we can apply if we need it later.

Unfortunately I couldn't dig much deeper into the differences aside from the aggregate change in instructions, since it looks like callgrind fudged some of the instruction attribution (`at::to_cpu` takes up a ton of instructions, but I don't see any attribution for the `at::native::add` kernel anywhere).

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#58065**

Differential Revision: [D28833085](https://our.internmc.facebook.com/intern/diff/D28833085)

### CODE REVIEW

**Comment by ezyang** [comment]
```
return at::native::call_fallback_fn<&xla_cpu_fallback, decltype(at::addmm)>::call("aten::addmm", self, mat1, mat2, beta, alpha);
```

It seems to me that if we could work out some way to template on `at::_ops::addmm` directly, then you could eliminate the necessity for the string op name too, because you should be able to index each of the disambiguated symbols to the relevant schema string.

**Comment by ezyang** [comment]
This is head in clouds stuff, but seeing how algorithmically similar the codegen and the boxed fallbacks are, I dream of a universe where we can write the semantics of a boxed fallback once, and then translate this into a codegen pass ala partial evaluation

File: aten/src/ATen/native/CPUFallback.cpp
Comment (ezyang) [inline aten/src/ATen/native/CPUFallback.cpp:61]: You don't have to be worried, vector constructor is

```

template <class T, IValue::enable_if_ivalue_constructible<T>>
inline IValue::IValue(const std::vector<T>& v) : IValue(c10::List<T>()) {
  auto list = to<c10::List<T>>();
  list.reserve(v.size());
  for (const auto& e : v) {
    list.push_back(e);
  }
}
```

your version is more efficient anyway tho!

File: aten/src/ATen/native/CPUFallback.cpp
Comment (ezyang) [inline aten/src/ATen/native/CPUFallback.cpp:63]: It's a little wonky here that you don't arrange for all tensors to be converted to CPU at once; e.g., if you have `f(Tensor, TensorList)`, first TensorList will get materialized, then Tensor.

File: aten/src/ATen/native/CPUFallback.cpp
Comment (ezyang) [inline aten/src/ATen/native/CPUFallback.cpp:101]: Wouldn't it be better to error in this case?

File: aten/src/ATen/native/CPUFallback.cpp
Comment (ezyang) [inline aten/src/ATen/native/CPUFallback.cpp:116]: 🚨 quadratic police 🚨

File: aten/src/ATen/native/CPUFallback.cpp
Comment (ezyang) [inline aten/src/ATen/native/CPUFallback.cpp:15]: Perhaps the stock `at::to_cpu` should just support undefined? IDK...

**Review by ezyang — APPROVED** [review]
This looks good. @zdevito, does 10-20% for working on boxed code make sense to you? I had a hypothesis that some of the cost was from traversing metadata but I don't see any flagrantly terrible logic on the metadata (there is a quadratic loop but it shouldn't matter).

A high level question I'd ask you is how you feel about the boxed fallback compared to the code generator pipeline. It seems like it's less code, how did you feel about the complexity? 

Other miscellanea:
* Higher priority boxed fallback seems like a reasonable thing to add.

File: aten/src/ATen/native/CPUFallback.h
Comment (bhosmer) [inline aten/src/ATen/native/CPUFallback.h:15]: Probably worth a brief comment describing the situation this is useful for.

File: aten/src/ATen/native/CPUFallback.h
Comment (bhosmer) [inline aten/src/ATen/native/CPUFallback.h:36]: Similarly, I'd add a comment here describing why you'd use this one instead of `call_fallback_fn`

**Review by bhosmer — APPROVED** [review]
Couple minor things but yeah LG! 

Re precedence vs `CompositeImplicitAutograd`, I definitely think we need to be careful though. Longer discussion.

File: aten/src/ATen/native/CPUFallback.cpp
Comment (bdhirsh) [reply aten/src/ATen/native/CPUFallback.cpp:15]: Agreed, I'll try adding that in. At some point I was debugging some failures related to undefined tensor and probably put the logic directly here out of frustration / to be extra safe 😛.

File: aten/src/ATen/native/CPUFallback.cpp
Comment (bdhirsh) [reply aten/src/ATen/native/CPUFallback.cpp:63]: > yeah I'd go further ask why we're collecting the individual tensor args into a temporary list here and passing it to the helper that has the extra list-related logic

This is actually needed for XLA. The individual xla tensor arguments can't be independently copied to CPU, there's some shared context between them involved with the copying. If you look at the xla implementation of `at::_to_cpu()`, it eventually [calls some fusing logic here](https://github.com/pytorch/xla/blob/280348439b23fd7aef5c57c9b92298f8890bcc98/torch_xla/csrc/tensor.cpp#L955).

> It's a little wonky here that you don't arrange for all tensors to be converted to CPU at once

Is the part you don't like the fact that arguments won't necessarily be copied to CPU in the order that they come in the operator's schema? That seems fair. But just to confirm, you're not suggesting a single `at::_to_cpu()` call with all of the Tensor + TensorList args?

File: aten/src/ATen/native/CPUFallback.cpp
Comment (bdhirsh) [reply aten/src/ATen/native/CPUFallback.cpp:66]: absolutely no significance, need to clean up some of my comments 😛 (that includes the "I'm worried" comment that Ed pointed out further up - definitely planned on removing that before eventually merging)

File: aten/src/ATen/native/CPUFallback.cpp
Comment (bdhirsh) [reply aten/src/ATen/native/CPUFallback.cpp:101]: Yeah, agreed; I left it like that to maintain the current behavior for XLA, but it's a bug that we should probably error on instead.

The only view op that it applies to for xla is `unfold`, since it's non-composite and doesn't have an XLA lowering:

```
>>> a = torch.arange(4)
>>> b = a.unfold(0, 2, 1)
>>> b[0] = -1
>>> a
tensor([-1, -1,  2,  3])
>>> a = torch.arange(4, device=xm.xla_device())
>>> b = a.unfold(0, 2, 1)
>>> b[0] = -1
>>> a
tensor([0, 1, 2, 3], device='xla:0')
```

File: aten/src/ATen/native/CPUFallback.cpp
Comment (bdhirsh) [reply aten/src/ATen/native/CPUFallback.cpp:116]: Yeah, this is totally something we could precompute. Right now we have a way to go from `Argument`->`AliasInfo`, but there's no direct link from `AliasInfo` -> [list_of_args_with_that_alias]. I could avoid the runtime quadratic by pre-computing that map in this fallback, but idk if it's worth the effort - it's not reaaaally quadratic since you only do the linear search for each mutable alias'd output argument (there's usually only one). And in xla's case, afaik most people want to call functional ops that don't have any mutable alias'd outputs.

On the topic of boxed fallback perf: there are probably a bunch of things we could pre-compute at registration time to speed this up, and boxed fallbacks in general. They would all increase the size of `FunctionSchema` though, so it'd probably be worth benchmarking to make sure we actually get a speedup.

Just a few ideas (not sure if these are all fruitful):
- [the one you said] Precompute the mapping between alias'd inputs/outputs, which would avoid the double for loop here. Right now you can identify them by checking if their alias_info objects are the same, but we could store direct references between the input/output args instead. I'm not sure whether or not we'd have to deal with weird schemas though, like `(Tensor(a!) t1, Tensor(a!) t2) -> Tensor(a!)`
- Precompute which arguments are tensors/tensorlists, since it's probably a pretty common task in a boxed fallback to iterate through tensor arguments.
- Also I see the FunctionSchema stores a `std::vector<Argument>` - it might be an easy win to make these `SmallVector`.

File: aten/src/ATen/native/CPUFallback.h
Comment (bdhirsh) [reply aten/src/ATen/native/CPUFallback.h:15]: agreed

File: aten/src/ATen/native/CPUFallback.h
Comment (bdhirsh) [reply aten/src/ATen/native/CPUFallback.h:36]: I should have preemptively commented these. Once https://github.com/pytorch/pytorch/pull/58092 lands I can kill the  version above, and use this nicer version (and rename `call_fallback_fn2` to `call_fallback_fn`). I talked with Ed and there's probably an even nicer version of this where you just call `at::native::call_fallback_fn<&xla_cpu_fallback, at::_ops::_adaptive_avg_pool3d>::call(self, output_size);`, and `at::_ops::_adaptive_avg_pool3d` has enough information to give you both the op schema type and the op's name + overload name, that way users of this API only have to pass in the arguments + the fallback that they want to call.

Right now I'm planning on just landing the PR independently from that change and doing a fix-up later.

**Comment by bdhirsh** [comment]
> This looks good. @zdevito, does 10-20% for working on boxed code make sense to you? I had a hypothesis that some of the cost was from traversing metadata but I don't see any flagrantly terrible logic on the metadata (there is a quadratic loop but it shouldn't matter).
> 
> A high level question I'd ask you is how you feel about the boxed fallback compared to the code generator pipeline. It seems like it's less code, how did you feel about the complexity?
> 
> Other miscellanea:
> 
> * Higher priority boxed fallback seems like a reasonable thing to add.

This feels to me like a pretty big complexity decrease- the codegen for CPU fallbacks was pretty convoluted and ugly. It comes at the cost of requiring a little more work from the backend though, since they now need to write their own boxed fallback. This isn't too bad though, since the helper functions we provide do 99% of the work - just needs to be well documented. The different ideas for improving perf of the fallback all come with their own additional complexity though, so that seems to me like a bridge we can cross later when/if we want to further improve perf.

File: aten/src/ATen/native/CPUFallback.cpp
Comment (ailzhang) [reply aten/src/ATen/native/CPUFallback.cpp:61]: Merging all tensor args into one giant list is actually the best for XLA since in case some tensors share a part of underlying graph that part can be materialized only once. ;) We probably didn't care too much about perf here so we indeed did it separately in the current codegen, but I think it can be changed. 
In this PR I think it's fine if we document this inefficiency in the comment, also note that "materializing together" is a requirement specific to xla only". ;)

File: aten/src/ATen/native/CPUFallback.cpp
Comment (ailzhang) [reply aten/src/ATen/native/CPUFallback.cpp:97]: Yup we should make it clear that XLA need to handle all view ops in the backend instead of relying on fallback. IIRC @JackCaoG tried lowering `unfold` before but it's not merged due to complexity and priority. 
(Erroring out here is definitely helpful).

**Review by ailzhang — APPROVED** [review]
Nice! Thanks @bdhirsh !

File: aten/src/ATen/native/CPUFallback.cpp
Comment (bdhirsh) [reply aten/src/ATen/native/CPUFallback.cpp:61]: I'm ok with adding a note: If the only factor is perf, there's also technically a tradeoff where eager backends that use the CPU fallback would get a small perf hit from the extra indexing logic to merge and slice up the list of tensors. But that's probably pretty mild, and I'm thinking of this as a knob we can use later to improve CPU fallback perf for XLA if we need to.

File: aten/src/ATen/native/CPUFallback.cpp
Comment (bdhirsh) [reply aten/src/ATen/native/CPUFallback.cpp:97]: Haven't updated the PR yet, but I just tested the error message locally:
```
>>> a = torch.arange(4, device=xm.xla_device())
>>> b = a.unfold(0, 2, 1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: The operator aten::unfold appears to be a view operator, but it has no implementation for the backend "xla:0". View operators don't support falling back to run on the CPU, since the tensor's storage cannot be shared across devices.
```

File: aten/src/ATen/native/CPUFallback.cpp
Comment (bdhirsh) [inline aten/src/ATen/native/CPUFallback.cpp:138]: (is it just me or is GitHub not letting me see the original comment thread from the outdated changes)

@ailzhang it looks like XLA doesn't have full coverage of `torch.as_strided` either. The existing kernel [falls back to CPU in most cases.](https://github.com/pytorch/xla/blob/280348439b23fd7aef5c57c9b92298f8890bcc98/torch_xla/csrc/aten_xla_type.cpp#L655). If that just seems like a coverage gap, I can file an issue (and not block the PR on that getting fixed, just need to modify some xla tests).
```
>>> t = torch.ones(4, device=xm.xla_device())
>>> t2 = t.as_strided((2,), (2,))
>>> t2[0] = -1
>>> t
tensor([1., 1., 1., 1.], device='xla:0') # should have been mutated
>>> t2
tensor([-1.,  1.], device='xla:0')
>>>
```

Comment (JackCaoG) [reply aten/src/ATen/native/CPUFallback.cpp:138]: I will strongly prefer not to error out in this case, reason stated in https://github.com/pytorch/xla/issues/2964#issuecomment-847274777

Comment (bdhirsh) [reply aten/src/ATen/native/CPUFallback.cpp:138]: Warning instead of erroring to avoid being BC-breaking seems reasonable to me.

Comment (bdhirsh) [reply aten/src/ATen/native/CPUFallback.cpp]: Tested out the warn:
```
>>> import torch
>>> import torch_xla.core.xla_model as xm
>>> a = torch.ones(4, device=xm.xla_device())
>>> b = a.as_strided((2,), (2,))
__main__:1: UserWarning: 0The operator aten::as_strided appears to be a view operator, but it has no implementation for the backend "xla:0". View operators don't support falling back to run on the CPU, since the tensor's storage cannot be shared across devices. (Triggered internally at  ../aten/src/ATen/native/CPUFallback.cpp:142.)
>>>
```

**Comment by bdhirsh** [comment]
@ezyang @bhosmer the description is a little out of date; I made changes mostly in `aten/src/ATen/native/CPUFallback.h` to the unboxed->boxed API, which uses the new `at::_ops` metadata.

**Comment by bdhirsh** [comment]
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28833085).

**Comment by bdhirsh** [comment]
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28833085).

**Comment by bdhirsh** [comment]
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28833085).

**Comment by bdhirsh** [comment]
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28833085).

**Comment by bdhirsh** [comment]
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28833085).

**Comment by bdhirsh** [comment]
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28833085).

**Comment by bdhirsh** [comment]
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28833085).

**Comment by bdhirsh** [comment]
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28833085).

**Comment by bdhirsh** [comment]
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28833085).

**Comment by bdhirsh** [comment]
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28833085).

**Pull request closed without merging**
